Ubuntu Focal build fails

### Qubes OS release

R 4.1



### Brief summary

The Ubuntu Focal template build fails with:

```
....
Removing 'local diversion of /sbin/initctl to /sbin/initctl.distrib'
INFO: Attempting to kill any processes still running in '/home/user/qubes-builder/qubes-src/linux-template-builder/mnt' before un-mounting
INFO: umount /home/user/qubes-builder/qubes-src/linux-template-builder/mnt/sys
INFO: Attempting to kill any processes still running in '/home/user/qubes-builder/qubes-src/linux-template-builder/mnt' before un-mounting
INFO: umount /home/user/qubes-builder/qubes-src/linux-template-builder/mnt/run
make[1]: *** [Makefile:69: rootimg-build] Error 1
```


### Steps to reproduce

- Follow https://github.com/Qubes-Community/Contents/blob/master/docs/building/building-archlinux-template.md
- In `make template` step the build fails.



### Expected behavior

The build succeeds.



### Actual behavior

`make[1]: *** [Makefile:69: rootimg-build] Error 1`

Same issue, tried this with 3 fresh VMs now. There isn't any information the logs or stdout as far as I can tell. Happy to look at whatever logs may be helpful.

in the setup script, there are two options for ubuntu, `focal` and `focal+desktop`. I didn't find any mention what the `+desktop` part was, so I only tried to build `focal`.

build-VM has 400-4000MB memory, 30GB storage, 2vcpus

The last few lines in `build-logs template-focal.log`:

```
[.... other packages that are already newest version]
xauth is already the newest version (1:1.1-0ubuntu1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Removing 'local diversion of /sbin/initctl to /sbin/initctl.distrib'
INFO: Attempting to kill any processes still running in '/home/user/qubes-builder/qubes-src/linux-template-builder/mnt' before un-mounting
INFO: umount /home/user/qubes-builder/qubes-src/linux-template-builder/mnt/sys
INFO: Attempting to kill any processes still running in '/home/user/qubes-builder/qubes-src/linux-template-builder/mnt' before un-mounting
INFO: umount /home/user/qubes-builder/qubes-src/linux-template-builder/mnt/run
make[1]: *** [Makefile:69: rootimg-build] Error 1
```

--- a/template_qubuntu/01_install_core.sh
+++ b/template_qubuntu/01_install_core.sh
@@ -1,13 +1,15 @@
 #!/bin/bash -e
 # vim: set ts=4 sw=4 sts=4 et :
 
-if [ "$VERBOSE" -ge 2 -o "$DEBUG" == "1" ]; then
+if [ "$DEBUG" == "1" ]; then
     set -x
 fi
 
 # Source external scripts
-source "${SCRIPTSDIR}/vars.sh"
-source "${SCRIPTSDIR}/distribution.sh"
+# shellcheck source=template_debian/vars.sh
+source "${TEMPLATE_CONTENT_DIR}/vars.sh"
+# shellcheck source=template_debian/distribution.sh
+source "${TEMPLATE_CONTENT_DIR}/distribution.sh"
 
 ##### '-------------------------------------------------------------------------
 debug ' Installing base system using debootstrap'
@@ -20,23 +22,24 @@ buildStep "${0}" "pre"
 
 
 bootstrap() {
-    for mirror in ${DEBIAN_MIRRORS[@]}; do
-        if [ ! -d "${INSTALLDIR}/${TMPDIR}" ]; then
-            mkdir -m 1777 -p "${INSTALLDIR}/${TMPDIR}"
+    for mirror in "${DEBIAN_MIRRORS[@]}"; do
+        if [ ! -d "${INSTALL_DIR}/${TMPDIR}" ]; then
+            # shellcheck disable=SC2174
+            mkdir -m 1777 -p "${INSTALL_DIR}/${TMPDIR}"
         fi
-        rm -rf "${INSTALLDIR}/${TMPDIR}/dummy-repo"
-        mkdir -p "${INSTALLDIR}/${TMPDIR}/dummy-repo/dists/${DIST}"
-        mkdir -p "${INSTALLDIR}/${TMPDIR}/dummy-repo/dists/${DIST}/main/binary-amd64"
-        mkdir -p "${INSTALLDIR}/${TMPDIR}/dummy-repo/dists/${DIST}/universe/binary-amd64"
-        echo ${mirror} > "${INSTALLDIR}/${TMPDIR}/.mirror"
+        rm -rf "${INSTALL_DIR}/${TMPDIR}/dummy-repo"
+        mkdir -p "${INSTALL_DIR}/${TMPDIR}/dummy-repo/dists/${DIST_CODENAME}"
+        mkdir -p "${INSTALL_DIR}/${TMPDIR}/dummy-repo/dists/${DIST_CODENAME}/main/binary-amd64"
+        mkdir -p "${INSTALL_DIR}/${TMPDIR}/dummy-repo/dists/${DIST_CODENAME}/universe/binary-amd64"
+        echo "${mirror}" > "${INSTALL_DIR}/${TMPDIR}/.mirror"
 
         mirror_no_proto=${mirror#*://}
         # depending on debootstrap version, Release files can be stored under
         # different names; this function needs _some_ correctly signed file for
         # dummy repository
         release_location_candidates=( \
-            "${INSTALLDIR}/var/lib/apt/lists/${mirror_no_proto//\//_}_dists_${DIST}_Release" \
-            "${INSTALLDIR}/var/lib/apt/lists/debootstrap.invalid_dists_${DIST}_Release" \
+            "${INSTALL_DIR}/var/lib/apt/lists/${mirror_no_proto//\//_}_dists_${DIST_CODENAME}_Release" \
+            "${INSTALL_DIR}/var/lib/apt/lists/debootstrap.invalid_dists_${DIST_CODENAME}_Release" \
         )
 
         apt_https_pkgs="apt-transport-https,ca-certificates"
@@ -44,46 +47,69 @@ bootstrap() {
         # them. Needs to copy Release{,.gpg} to a dummy _local_ repo, because
         # debootstrap insists on downloading it each time but we want to be sure to use
         # packages downloaded earlier (and logged)
-        COMPONENTS="" $DEBOOTSTRAP_PREFIX debootstrap \
+        # shellcheck disable=SC2154
+        COMPONENTS="" "${DEBOOTSTRAP_PREFIX[@]}" debootstrap \
             --arch=amd64 \
             --include="ncurses-term,locales,tasksel,$apt_https_pkgs,$eatmydata_maybe" \
             --components=main,universe \
             --download-only \
-            --keyring="${SCRIPTSDIR}/../keys/${DIST}-${DISTRIBUTION}-archive-keyring.gpg" \
-            "${DIST}" "${INSTALLDIR}" "${mirror}" && \
-        sha256sum "${INSTALLDIR}/var/cache/apt/archives"/*.deb && \
+            --keyring="${KEYS_DIR}/${DIST_CODENAME}-${DIST_NAME}-archive-keyring.gpg" \
+            "${DIST_CODENAME}" "${INSTALL_DIR}" "${mirror}" && \
+        sha256sum "${INSTALL_DIR}/var/cache/apt/archives"/*.deb && \
         for release_location in "${release_location_candidates[@]}"; do
             if [ -r "${release_location}" ]; then
                 cp "${release_location}" \
-                    "${INSTALLDIR}/${TMPDIR}/dummy-repo/dists/${DIST}/Release"
-                cp "${release_location}.gpg" \
-                    "${INSTALLDIR}/${TMPDIR}/dummy-repo/dists/${DIST}/Release.gpg"
+                    "${INSTALL_DIR}/${TMPDIR}/dummy-repo/dists/${DIST_CODENAME}/Release"
+                inrelease_location="${release_location%Release}InRelease"
+                if [ -r "$inrelease_location" ]; then
+                    cp "${inrelease_location}" \
+                        "${INSTALL_DIR}/${TMPDIR}/dummy-repo/dists/${DIST_CODENAME}/InRelease"
+                fi
+                if [ -r "${release_location}.gpg" ]; then
+                    cp "${release_location}.gpg" \
+                        "${INSTALL_DIR}/${TMPDIR}/dummy-repo/dists/${DIST_CODENAME}/Release.gpg"
+                fi
                 cp "${release_location%_Release}_main_binary-amd64_Packages" \
-                    "${INSTALLDIR}/${TMPDIR}/dummy-repo/dists/${DIST}/main/binary-amd64/Packages"
+                    "${INSTALL_DIR}/${TMPDIR}/dummy-repo/dists/${DIST_CODENAME}/main/binary-amd64/Packages"
                 cp "${release_location%_Release}_universe_binary-amd64_Packages" \
-                    "${INSTALLDIR}/${TMPDIR}/dummy-repo/dists/${DIST}/universe/binary-amd64/Packages"
+                    "${INSTALLDIR}/${TMPDIR}/dummy-repo/dists/${DIST_CODENAME}/universe/binary-amd64/Packages"
                 break
             fi
         done && \
-        COMPONENTS="" $DEBOOTSTRAP_PREFIX debootstrap \
+        COMPONENTS="" "${DEBOOTSTRAP_PREFIX[@]}" debootstrap \
             --arch=amd64 \
             --include="ncurses-term,locales,tasksel,$apt_https_pkgs,$eatmydata_maybe" \
             --components=main,universe \
-            --keyring="${SCRIPTSDIR}/../keys/${DIST}-${DISTRIBUTION}-archive-keyring.gpg" \
-            "${DIST}" "${INSTALLDIR}" "file://${INSTALLDIR}/${TMPDIR}/dummy-repo" && \
-        echo "deb ${mirror} ${DIST} main universe" > ${INSTALLDIR}/etc/apt/sources.list && \
+            --keyring="${KEYS_DIR}/${DIST_CODENAME}-${DIST_NAME}-archive-keyring.gpg" \
+            "${DIST_CODENAME}" "${INSTALL_DIR}" "file://${INSTALL_DIR}/${TMPDIR}/dummy-repo" && \
+        echo "deb ${mirror} ${DIST_CODENAME} main" > "${INSTALL_DIR}"/etc/apt/sources.list && \
         return 0
     done
-    error "Debootstrap failed!"
-    exit 1;
+    return 1
 }
 
 
-if ! [ -f "${INSTALLDIR}/${TMPDIR}/.prepared_debootstrap" ]; then
+if ! [ -f "${INSTALL_DIR}/${TMPDIR}/.prepared_debootstrap" ]; then
     #### "------------------------------------------------------------------
-    info " $(templateName): Installing base '${DISTRIBUTION}-${DIST}' system"
+    info " $(templateName): Installing base '${DIST_NAME}-${DIST_CODENAME}' system"
     #### "------------------------------------------------------------------
-    bootstrap || exit 1
+    retry=0
+    while ! bootstrap
+    do
+        if [ $retry -le 3 ]; then
+            echo "Error in debootstrap. Sleeping 60 sec before retrying..."
+            retry=$((retry + 1))
+            sleep 60
+        else
+            echo "Error in debootstrap. Aborting due to too much retries."
+            exit 1
+        fi
+    done
+
+    #### '----------------------------------------------------------------------
+    info ' Download APT metadata'
+    #### '----------------------------------------------------------------------
+    chroot_cmd apt-get update || exit 1
 
     #### '----------------------------------------------------------------------
     info ' Configure keyboard'
@@ -103,10 +129,11 @@ if ! [ -f "${INSTALLDIR}/${TMPDIR}/.prepared_debootstrap" ]; then
 
     # TMPDIR is set in vars.  /tmp should not be used since it will be cleared
     # if building template with LXC contaniners on a reboot
-    mkdir -m 1777 -p "${INSTALLDIR}/${TMPDIR}"
+    # shellcheck disable=SC2174
+    mkdir -m 1777 -p "${INSTALL_DIR}/${TMPDIR}"
 
     # Mark section as complete
-    touch "${INSTALLDIR}/${TMPDIR}/.prepared_debootstrap"
+    touch "${INSTALL_DIR}/${TMPDIR}/.prepared_debootstrap"
 
     # If SNAPSHOT=1, Create a snapshot of the already debootstraped image
     createSnapshot "debootstrap"

--- a/template_qubuntu/vars.sh
+++ b/template_qubuntu/vars.sh
@@ -1,7 +1,24 @@
 #!/bin/bash -e
 # vim: set ts=4 sw=4 sts=4 et :
 
-source ./functions.sh
+# shellcheck disable=SC2034
+
+#
+# Handle legacy builder
+#
+
+if [ "0${IS_LEGACY_BUILDER}" -eq 1 ]; then
+    TEMPLATE_SCRIPTS_DIR="$(readlink -f .)"
+    KEYS_DIR="$SCRIPTSDIR/../keys"
+    DIST_CODENAME="$DIST"
+    DIST_NAME="$DISTRIBUTION"
+    PACKAGES_DIR="$BUILDER_REPO_DIR"
+    # shellcheck disable=SC2153
+    INSTALL_DIR="${INSTALLDIR}"
+fi
+
+# shellcheck disable=SC1091
+source "${TEMPLATE_SCRIPTS_DIR}/functions.sh" >/dev/null
 
 # ==============================================================================
 # Global variables and functions
@@ -16,33 +33,46 @@ export PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
 TMPDIR="/tmp"
 
 # ------------------------------------------------------------------------------
+# The codename of the Ubuntu version to install.
+# jessie = testing, wheezy = stable
+# ------------------------------------------------------------------------------
+DEBIANVERSION=${DIST_CODENAME}
+
+# ------------------------------------------------------------------------------
 # Location to grab Ubuntu packages
 # ------------------------------------------------------------------------------
-DEBIAN_MIRRORS=(
+
+DEFAULT_DEBIAN_MIRRORS=(
     'http://archive.ubuntu.com/ubuntu'
 )
+
+# DEBIAN_MIRRORS can be set in configuration file to override the defaults
+if [ -z "${DEBIAN_MIRRORS}" ]; then
+    read -r -a DEBIAN_MIRRORS<<<"${DEFAULT_DEBIAN_MIRRORS[@]}"
+fi
 
 # ------------------------------------------------------------------------------
 # Kernel package to install
 # ------------------------------------------------------------------------------
-KERNEL_PACKAGE_NAME="linux-image-generic"
+KERNEL_PACKAGE_NAME="linux-image-amd64"
 
 # ------------------------------------------------------------------------------
 # apt-get configuration options
 # ------------------------------------------------------------------------------
-APT_GET_OPTIONS="-o Dpkg::Options::=--force-confnew -o Dpkg::Options::=--force-unsafe-io --yes"
-APT_GET_OPTIONS+=" -o Acquire::Retries=3"
-if [ -n "$REPO_PROXY" ]; then
-     APT_GET_OPTIONS+=" -o Acquire::http::Proxy=${REPO_PROXY}"
-     DEBOOTSTRAP_PREFIX+=" env http_proxy=${REPO_PROXY}"
-fi
-
+APT_GET_OPTIONS=("-o" "Dpkg::Options::=--force-confnew" "--yes")
+APT_GET_OPTIONS+=("-o" "Acquire::Retries=3")
 
 if [ "0${BUILDER_TURBO_MODE}" -gt 0 ]; then
-    APT_GET_OPTIONS+=" -o Dpkg::Options::=--force-unsafe-io"
+    APT_GET_OPTIONS+=("-o" "Dpkg::Options::=--force-unsafe-io")
     eatmydata_maybe=eatmydata
 fi
 
-containsFlavor 'no-recommends' && {
-    APT_GET_OPTIONS+=" -o APT::Install-Recommends=0  -o APT::Install-Suggests=0" 
-} || true
+if [ -n "$REPO_PROXY" ]; then
+    APT_GET_OPTIONS+=("-o" "Acquire::http::Proxy=${REPO_PROXY}")
+    DEBOOTSTRAP_PREFIX+=("env" "http_proxy=${REPO_PROXY}")
+    DEBOOTSTRAP_PREFIX+=("env" "https_proxy=${REPO_PROXY}")
+fi
+
+if containsFlavor 'no-recommends'; then
+    APT_GET_OPTIONS+=("-o" "APT::Install-Recommends=0" "-o" "APT::Install-Suggests=0")
+fi

